src/measure_theory/pi_system.lean

```diff
@@ -0,0 +1,170 @@
+
+lemma generate_from_generate_pi_system_eq {α} {g : set (set α)} :
+  generate_from (generate_pi_system g) = generate_from g :=
+begin
+  apply le_antisymm; apply generate_from_le,
+  { exact λ t h_t, generate_from_measurable_set_of_generate_pi_system t h_t, },
+  { exact λ t h_t, measurable_set_generate_from (generate_pi_system.base h_t), },
+end
+
+/- This theorem shows that every element of the pi system generated by the union of the
+   pi systems can be represented by a finite union of elements from the pi systems. -/
+lemma mem_generate_pi_system_Union_elim {α β} {g : β → set (set α)}
+  (h_pi : ∀ b, is_pi_system (g b)) (t : set α) (h_t : t ∈ generate_pi_system (⋃ b, g b)) :
+  (∃ (T : finset β) (f : β → set α), (t = ⋂ b ∈ T, f b) ∧ (∀ b ∈ T, f b ∈ (g b))) :=
+begin
+  induction h_t with s h_s s t' h_gen_s h_gen_t' h_nonempty h_s h_t',
+  { rcases h_s with ⟨t', ⟨⟨b, rfl⟩, h_s_in_t'⟩⟩,
+    refine ⟨{b}, (λ _, s), _⟩,
+    simpa using h_s_in_t', },
+  { rcases h_t' with ⟨T_t', ⟨f_t', ⟨rfl, h_t'⟩⟩⟩,
+    rcases h_s with ⟨T_s, ⟨f_s, ⟨rfl, h_s⟩ ⟩ ⟩,
+    use [(T_s ∪ T_t'), (λ (b:β),
+      if (b ∈ T_s) then (if (b ∈ T_t') then (f_s b ∩ (f_t' b)) else (f_s b))
+      else (if (b ∈ T_t') then (f_t' b) else (∅ : set α)))],
+    split,
+    { ext a,
+      simp_rw [set.mem_inter_iff, set.mem_Inter, finset.mem_union, or_imp_distrib],
+      rw ← forall_and_distrib,
+      split; intros h1 b; by_cases hbs : b ∈ T_s; by_cases hbt : b ∈ T_t'; specialize h1 b;
+        simp only [hbs, hbt, if_true, if_false, true_implies_iff, and_self, false_implies_iff,
+          and_true, true_and] at h1 ⊢,
+      all_goals { exact h1, }, },
+    intros b h_b,
+    split_ifs with hbs hbt hbt,
+    { refine h_pi b (f_s b) (f_t' b) (h_s b hbs) (h_t' b hbt) (set.nonempty.mono _ h_nonempty),
+      exact set.inter_subset_inter (set.bInter_subset_of_mem hbs) (set.bInter_subset_of_mem hbt), },
+    { exact h_s b hbs, },
+    { exact h_t' b hbt, },
+    { rw finset.mem_union at h_b,
+      apply false.elim (h_b.elim hbs hbt), }, },
+end
+
+/- This is similar to mem_generate_pi_system_Union_elim', but
+   focuses on a set of elements in b, as opposed to the whole type. -/
+lemma mem_generate_pi_system_Union_elim' {α β} {g : β → set (set α)} {s: set β}
+  (h_pi : ∀ b ∈ s, is_pi_system (g b)) (t : set α) (h_t : t ∈ generate_pi_system (⋃ b ∈ s, g b)) :
+  (∃ (T : finset β) (f : β → set α), (↑T ⊆ s) ∧ (t = ⋂ b ∈ T, f b) ∧ (∀ b ∈ T, f b ∈ (g b))) :=
+begin
+  classical,
+  have h1 := @mem_generate_pi_system_Union_elim α (subtype s) (g ∘ subtype.val) _ t _,
+  rcases h1 with ⟨T, ⟨f,⟨ rfl, h_t'⟩ ⟩⟩,
+  use (T.image subtype.val),
+  use (function.extend subtype.val f (λ (b:β), (∅ : set α))),
+  split,
+  { simp },
+  split,
+  { ext a, split;
+    { simp only [set.mem_Inter, subtype.forall, finset.set_bInter_finset_image],
+      intros h1 b h_b h_b_in_T,
+      have h2 := h1 b h_b h_b_in_T, revert h2,
+      rw function.extend_apply subtype.val_injective, apply id } },
```

Comment:
It is better to put only one tactic per line, unless you close the whole goal in one line, like you do at line 165.
```suggestion
      have h2 := h1 b h_b h_b_in_T,
      revert h2,
      rw function.extend_apply subtype.val_injective,
      apply id } },
```